protocol/lavasession/consumer_session_manager.go

```diff
@@ -356,32 +394,71 @@ func (csm *ConsumerSessionManager) getValidProviderAddress(ignoredProvidersList
 	}
 
 	providers := csm.providerOptimizer.ChooseProvider(csm.validAddresses, ignoredProvidersList, cu, requestedBlock, OptimizerPerturbation)
-	if len(providers) != 1 {
-		return "", utils.LavaFormatError("could not choose a provider", nil, utils.Attribute{Key: "Provider list", Value: csm.validAddresses}, utils.Attribute{Key: "IgnoredProviderList", Value: ignoredProvidersList})
-	}
-	return providers[0], nil
+	return providers, nil
 }
 
-func (csm *ConsumerSessionManager) getValidConsumerSessionsWithProvider(ignoredProviders *ignoredProviders, cuNeededForSession uint64, requestedBlock int64) (consumerSessionsWithProvider *ConsumerSessionsWithProvider, providerAddress string, currentEpoch uint64, err error) {
+func (csm *ConsumerSessionManager) getValidConsumerSessionsWithProvider(ignoredProviders *ignoredProviders, cuNeededForSession uint64, requestedBlock int64) (sessionWithProviderMap SessionWithProviderMap, err error) {
 	csm.lock.RLock()
 	defer csm.lock.RUnlock()
-	currentEpoch = csm.atomicReadCurrentEpoch() // reading the epoch here while locked, to get the epoch of the pairing.
+	currentEpoch := csm.atomicReadCurrentEpoch() // reading the epoch here while locked, to get the epoch of the pairing.
 	if ignoredProviders.currentEpoch < currentEpoch {
 		utils.LavaFormatDebug("ignoredProviders epoch is not the current epoch, resetting ignoredProviders", utils.Attribute{Key: "ignoredProvidersEpoch", Value: ignoredProviders.currentEpoch}, utils.Attribute{Key: "currentEpoch", Value: currentEpoch})
 		ignoredProviders.providers = make(map[string]struct{}) // reset the old providers as epochs changed so we have a new pairing list.
 		ignoredProviders.currentEpoch = currentEpoch
 	}
 
-	providerAddress, err = csm.getValidProviderAddress(ignoredProviders.providers, cuNeededForSession, requestedBlock)
+	// Fetch provider addresses
+	providerAddresses, err := csm.getValidProviderAddresses(ignoredProviders.providers, cuNeededForSession, requestedBlock)
```

Comment:
I stumbled on the issue with the approach with optional params, that's why I decided not to add it. Let me explain:

The biggest challenge with the current implementation (one we used so far) is that we are validating if the provider is valid in multiple places. 

So what do we have:
1. **getValidProviderAddresses** which returns an array of providers using optimizer (chooseProvider) or  PairingListEmptyError
2. **getValidConsumerSessionsWithProvider** which calls getValidProviderAddresses and before returning providers it validates providers https://github.com/lavanet/lava/blob/main/protocol/lavasession/consumer_session_manager.go#L382
3. **GetSessions** which calls getValidConsumerSessionsWithProvider and also before returning validates providers on many places (for example https://github.com/lavanet/lava/blob/main/protocol/lavasession/consumer_session_manager.go#L295) 

Our goal is to return as many providers as possible from what we initialized in the optimizer.

Let's focus on the **getValidConsumerSessionsWithProvider** method to analyze the problem:

The situation can be like this. Optimizer is initialized to return 5 providers. Whenever we call **getValidProviderAddresses** optimized will return either up to 5 providers or PairingListEmptyError. We call getValidProviderAddresses and it returns 5 but out of these 5, only 3 are valid. So we need to ping again and try to find the remaining 2. If we would use optional params we would ask again optimizer to return 2, but nothing is guaranteeing that these 2 will pass this validation https://github.com/lavanet/lava/blob/main/protocol/lavasession/consumer_session_manager.go#L382. With this implementation, if we see we are missing 2 providers out of the initial 5, we would ask the optimizer again to return providers, we would again iterate over the list and if at any moment we have the wanted number of providers we will return. https://github.com/lavanet/lava/blob/PRT-514-quorum-select-first-provider/protocol/lavasession/consumer_session_manager.go#L443. If again there no valid, we will ask the optimizer for more and so on till we either have wanted number or we have no more to check. 

I do not see how the optional param in **chooseProvider** method would simplify the logic, or how we can bypass the validations in both the **getValidConsumerSessionsWithProvider** and **GetSessions**

On the other hand, optional parameters in go are not implemented by design. Ofc there are ways to create optional param behavior but I am trying to avoid it as much as possible. Few which are familiar to me: 
1. **Variadic Functions**
```
func sum(test string, nums ...int) {

	fmt.Println(test)
	if len(nums) != 0 {
		fmt.Println(nums[0])
	}
}

func main() {
	sum("a", 1)
	sum("a")
}
```
The issue I see with them is that I am not enforcing how many arguments can exist. So someone could call sum("a", 1,2,4) and the function will still look good. Ofc we could handle all of this internally but compiler would not catch the issue

2. We could use structs
```
type Params struct {
  a, b, c int
}

func doIt(p Params) int {
  return p.a + p.b + p.c 
}

// you can call it without specifying all parameters
doIt(Params{a: 1, c: 9})
```

3. Functional Options
```
func NewConfig(options ...OptionFunc) *Config {
	config := &Config{}
	for _, option := range options {
		option(config)
	}
	return config
}
```

Can you please explain how we can utilize optional params to make this flow more optimal 🙏